eventsources/sources/azureservicebus/start.go

```diff
@@ -0,0 +1,182 @@
+)
+
+// StartListening starts listening events
+func (el *EventListener) StartListening(ctx context.Context, dispatch func([]byte, ...eventsourcecommon.Option) error) error {
+	log := logging.FromContext(ctx).
+		With(logging.LabelEventSourceType, el.GetEventSourceType(), logging.LabelEventName, el.GetEventName())
+
+	log.Info("started processing the Azure Service Bus event source...")
+	defer sources.Recover(el.GetEventName())
+
+	servicebusEventSource := &el.AzureServiceBusEventSource
+	connStr, err := common.GetSecretFromVolume(servicebusEventSource.ConnectionString)
+	if err != nil {
+		log.With("connection-string", servicebusEventSource.ConnectionString.Name).Errorw("failed to retrieve connection string from secret", zap.Error(err))
+		return err
+	}
+
+	log.Info("connecting to the service bus...")
+	clientOptions := servicebus.ClientOptions{}
+	if servicebusEventSource.TLS != nil {
+		tlsConfig, err := common.GetTLSConfig(servicebusEventSource.TLS)
+		if err != nil {
+			log.Errorw("failed to get the tls configuration", zap.Error(err))
+			return err
+		}
+		clientOptions.TLSConfig = tlsConfig
+	}
+
+	client, err := servicebus.NewClientFromConnectionString(connStr, &clientOptions)
+	if err != nil {
+		log.Errorw("failed to create a service bus client", zap.Error(err))
+		return err
+	}
+
+	var receiver *servicebus.Receiver
+	var receiverType ReceiverType
+
+	if servicebusEventSource.QueueName != "" {
+		log.Info("creating a queue receiver...")
+		receiverType = ReceiverTypeQueue
+		receiver, err = client.NewReceiverForQueue(servicebusEventSource.QueueName, &servicebus.ReceiverOptions{
+			ReceiveMode: servicebus.ReceiveModeReceiveAndDelete,
+		})
+	} else {
+		log.Info("creating a subscription receiver...")
+		receiverType = ReceiverTypeSubscription
+		receiver, err = client.NewReceiverForSubscription(servicebusEventSource.TopicName, servicebusEventSource.SubscriptionName, &servicebus.ReceiverOptions{
+			ReceiveMode: servicebus.ReceiveModeReceiveAndDelete,
+		})
+	}
+	if err != nil {
+		if receiverType == ReceiverTypeQueue {
+			log.With("queue", servicebusEventSource.QueueName).Errorw("failed to create a queue receiver", zap.Error(err))
+		} else {
+			log.With("topic", servicebusEventSource.TopicName, "subscription", servicebusEventSource.SubscriptionName).Errorw("failed to create a receiver for subscription", zap.Error(err))
+		}
+		return err
+	}
+
+	if servicebusEventSource.JSONBody {
+		log.Info("assuming all events have a json body...")
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("context has been cancelled, stopping the Azure Service Bus event source...")
+			if err := receiver.Close(ctx); err != nil {
+				log.Errorw("failed to close the receiver", zap.Error(err))
+				return err
+			}
+			return nil
+		default:
+			messages, err := receiver.ReceiveMessages(ctx, 1, nil)
+			if err != nil {
+				log.Errorw("failed to receive messages", zap.Error(err))
+				return err
```

Comment:
Returning error means quit it, is it expected?

Reply:
I would depend on the error it returns actually. It would be the same as https://github.com/argoproj/argo-events/blob/master/eventsources/sources/amqp/start.go#L280 fails

What happens when it quit? Will it restart automatically after?

Reply:
It will retry several times, if still fails in the end, the pod will crash and k8s will restart the pod, this is based on the assumption that there's only one event configured in the eventsource.

If there are multiple events configured in the eventsource object, and it fails after retry, it then stops retrying. But the pod is still running to serve the traffic for other events configuration.

My question is, could this error happen after some successful runs (e.g. due to temporary network issue), and it will recover? If that's the case, I suggest to let it `continue` and just print out error logs.

Reply:
I set it to `continue` now. As most of those errors would be recoverable from my understanding